code/ui/hud/chatbox/Chat.razor

```diff
@@ -0,0 +1,256 @@
+		max-height: 435px;
+		font-family: SourceSansPro-SemiBold;
+		text-stroke: 4px rgba(32,32,32,0.7);
+		font-size: 16px;
+		pointer-events: none;
+		transition: all 0.1s ease, border 0s ease;
+		overflow: hidden;
+		color: white;
+
+		textentry {
+			transition: opacity 0.1s ease;
+			flex-shrink: 0;
+			min-height: 44px;
+			opacity: 0;
+			background-color: rgba(32,32,32,0.7);
+			align-items: center;
+			padding: 8px;
+		}
+
+		.placeholder {
+			color: #c0cee7;
+		}
+
+		.content-label {
+			margin-left: 1px;
+		}
+
+		.placeholder, .content-label {
+			font-size: 18px;
+			font-family: SourceSansPro-SemiBold;
+			text-stroke: 4px rgba(32,32,32,0.7);
+			font-weight: 400;
+		}
+
+		.canvas {
+			flex-direction: column;
+			align-items: flex-start;
+			overflow: scroll;
+
+			ChatRow {
+				align-items: center;
+				max-width: 100%;
+				color: white;
+				opacity: 1;
+				flex-shrink: 0;
+				padding: 8px 10px;
+				transition: opacity 0.15s ease;
+
+				.header {
+					flex-shrink: 0;
+					flex-grow: 0;
+					padding-left: 4px;
+					font-weight: 800;
+					font-size: 25px;
+					padding-right: 8px;
+				}
+
+				.msg {
+					color: white;
+					font-size: 25px;
+				}
+
+				&.info {
+					.header, .msg {
+						color: darken( #c0cee7, 20% );
+					}
+				}
+
+				img {
+					flex-shrink: 0;
+					height: 32px;
+					width: 32px;
+					border-radius: 100px;
+				}
+
+				.name {
+					align-items: center;
+					justify-content: center;
+					border-radius: 8px;
+					padding: 1px;
+					padding-right: 10px;
+					color: #f3ff82;
+				}
+
+				&.noname {
+					.name {
+						display: none;
+					}
+				}
+
+				&.stale {
+					transition: opacity 1s ease;
+					opacity: 0;
+				}
+			}
+		}
+
+		&.open {
+			pointer-events: all;
+			backdrop-filter-blur: 10px;
+			background-color: rgba(16,16,16,0.4);
+			left: 32px;
+			border-radius: $rounding;
+
+			textentry {
+				opacity: 1;
+				width: 100%;
+				cursor: text;
+			}
+
+			.canvas {
+				ChatRow {
+					transition: opacity 0.1s ease;
+					opacity: 1;
+				}
+			}
+		}
+	}
+</style>
+
+<root>
+	<div class="canvas" @ref=Canvas />
+	<TextEntry @ref=Input onsubmit=@Submit onblur=@Close />
+</root>
+
+@code
+{
+	public static Chat Current;
+
+	public Panel Canvas { get; protected set; }
+	public TextEntry Input { get; protected set; }
+
+	Queue<ChatRow> Rows = new();
+
+	protected int MaxItems => 100;
+	protected float MessageLifetime => 10f;
+
+	public bool IsOpen
+	{
+		get => HasClass( "open" );
+		set
+		{
+			SetClass( "open", value );
+			if ( value )
+			{
+				Input.Focus();
+				Input.Text = string.Empty;
+				Input.Label.SetCaretPosition( 0 );
+			}
+		}
+	}
+
+	protected override void OnAfterTreeRender( bool firstTime )
+	{
+		base.OnAfterTreeRender( firstTime );
+
+		Canvas.PreferScrollToBottom = true;
+		Input.AcceptsFocus = true;
+		Input.AllowEmojiReplace = true;
+
+		Current = this;
+	}
+
+	public override void Tick()
+	{
+		if ( Sandbox.Input.Pressed( "Chat" ) )
```

Comment:
The pattern we've been following in Cinema so far is that any input code should happen inside of the Player Pawn class, which can then communicate to UI elements to tell them to open or close.

Generally UI elements should only ever **read** from player/pawn code, and should **never** write to player pawn code.

If you're trying to create a system that tracks whether the player pawn has the chat window open or not (/ is typing or not) that logic should be encapsulated inside of the Player Pawn code.

An idea on how to make this code conform to this pattern would be:
- In the player pawn code: When "Chat" is pressed, toggle `CurrentlyTyping`
- In the chat ui code: Read `Player.CurrentlyTyping`, and if it's true, we're open, otherwise we're closed

You could track the last frame state of `player.CurrentlyTyping` in the chat ui code if you want to do `onOpen/Close` events.

Reply:
My bad, I have mis-documented the CurrentlyTyping **string**. It's what the user is typing in the chatbox, but hasn't sent it yet. Those changes will reflect in the next commit